### DISCUSSION TITLE
Using `@cloudflare/worker-types` for the `app.d.ts`

### DISCUSSION THREAD

**Discussion by jokull** [sveltejs/kit#4898] [Q&A] [answered]
Is it possible to jack [something](https://github.com/cloudflare/workers-types/blob/18ca7fca0e341ef6a337594207c5a65aac75e2d4/manual-ts/pages.d.ts#L3) from `@cloudflare/worker-types` to have the `platform` variable be typed in SvelteKit hooks?

There's a [comment here](https://github.com/sveltejs/kit/blob/master/packages/kit/types/ambient.d.ts):

> It's possible to tell SvelteKit how to type objects inside your app by declaring the `App` namespace. By default, a new project will have a file called `src/app.d.ts` containing the following:

> By populating these interfaces, you will gain type safety when using `event.locals`, `event.platform`, `session` and `stuff`:

I'm looking for an example of this somewhere. The env is always a generic in `@cloudflare/worker-types`, so maybe you have to declare the type for each expected variable, like `ASSETS` in the case of Pages?

Any pointers appreciated.

**Comment by colinhowells** [comment, marked as answer]
_Edited to reflect current best practices_

Hopefully this gives you a pointer to solving your case, this is what I'm doing at the moment:

```ts
// /src/app.d.ts
declare global {
	interface Window {
		someObnoxiousGlobalObjectFromAdVendor: any;
	}
	namespace App {
		interface Error {
			code?: string; // we always have 'message', but adding code
		}
		interface Locals {
			user?: import('$lib/types').User; // in case you have users
		}
		interface PageData {
			someDataYouLike?: import('$lib/types').LikedData; // $page.data.someDataYouLike
		}
		interface Platform {
			cf?: IncomingRequestCfProperties;
			env?: {
				CF_PAGES_COMMIT_SHA: string; // I use this for Kit version
				SOME_KV_BINDING: KVNamespace;
			};
		}
	}
}

export {};
```

**Reply by jokull** [reply]
Mate, thank you! Makes sense!

**Comment by xpat** [comment]
@jokull @colinhowells Mind if I ask why `.src/global.d.ts` and not `app.d.ts`?

I'm trying out Cloudflare's D1 database.  There's an [example app](https://github.com/elithrar/svelte-demo-d1/tree/main) that I've only managed to get working by creating the Sveltekit project with TypeScript.  It didn't work if I created it with JsDoc.  If I create a new SK project with just javascript, there is no `app.d.ts` so I'm wondering if it's only possible to run that Cloudflare app with a SK TypeScript.  That's the bigger question for me, but I'm also curious as to why `global.d.ts` and not `app.d.ts`.

```
// See https://kit.svelte.dev/docs/types#app
// for information about these interfaces
declare global {
        namespace App {
                // interface Error {}
                // interface Locals {}
                // interface PageData {}
                   interface Platform {
      env: {
        DB: D1Database;
      };
      context: {
        waitUntil(promise: Promise<any>): void;
      };
      caches: Cache
        }
}

export {};
```

**Reply by colinhowells** [reply]
Ha yeah you are absolutely right, my post is pretty old now and `app.d.ts` is definitely the one you want to use (I've edited the original post).

If you have random npm packages, you can use `src/ambient.d.ts` for those

```ts
// src/ambient.d.ts, just an example of things I put in there
declare module '@beyonk/svelte-google-analytics';
declare module '@rollup/plugin-graphql';
declare module '@sentry/vite-plugin';
declare module 'toucan-js';
declare module '*.gql';
```

and if you have types of your own then `types.d.ts` in `/lib` allows you to do `/** @type {import('$lib/types').Thing} */`

```ts
// /src/lib/types.d.ts
type SomeType = {
	id: string;
	text: string;
	excellent: boolean;
};
type SomeTypes = SomeType[];
export type Thing = 'hamsters' | 'kumquats' | 'cabbage'; // this type is importable cos 'export'
``

**Comment by axel-rock** [comment]
I'm jumping in this pretty old thread already, I'm facing an issue in SvelteKit.

I've been doing the following:

```
import type { Session, SupabaseClient, User as AuthUser } from '@supabase/supabase-js'
import type { Database } from '$lib/db/types'
declare global {
	namespace App {
		interface Locals {
			supabase: SupabaseClient<Database>
			safeGetSession: () => Promise<{ session: Session | null; user: AuthUser | null }>
		}
		interface PageData {
			session: Session | null
		}
	}

	type Message = Database['public']['Tables']['messages']['Row']
}

export {}
```

Using the Message type from Database, to make it available without import, similar to what an `ambient.d.ts` file would do.

My life has been great, until today: I installed Sentry, and Sentry depends on `@cloudlare/worker-types`, which seems to export a global type called Message, and takes priority over this, instantly making my entire codebase's types wrong.

I tried moving my Message type into an ambient file, or within my lib folder as .d.ts file, none of these option work.

Am I missing a way to avoid this type contamination?

**Reply by jokull** [reply]
I have a hack for this using pnpm overrides if you are desperate

**Reply by axel-rock** [reply]
@jokull I am very much interested! I googled that, but not very deep, no success so far

**Reply by jokull** [reply]
I have a pnpm workspace - top level package.json:

```json5
{
    // ...
    "overrides": {
      "@cloudflare/workers-types": "link:./packages/empty-types"
    }
  }
}
```

then the empty-types package

```json5
{
  "name": "@empty/types",
  "version": "1.0.0"
}
```

**Reply by colinhowells** [reply]
Couldn't you get around this by putting your Message type somewhere other than global, and declaratively importing it each time? Alternatively you could rename your Message so the names didn't conflict. That way you wouldn't lose the CF types, either.

I don't know your circumstances so I'm sure you have your reasons, just curious. I might run into something like this myself later.

**Reply by jokull** [reply]
Didn't work for me. If a 3rd party package depended on @cloudflare-workers-types (which many still do) then it would be pulled into the globals, and I couldn't give generated types precedence.

**Reply by teemingc** [reply]
@lukaszpolowczyk is there a way to prevent the CF types from polluting the global namespace when users install Sentry?

I noticed I get the global types too when I install the Drizzle D1 adapter.

However, the recommended way to get CF types for users is to generate the types with Wrangler https://developers.cloudflare.com/workers/languages/typescript/#generate-types

**Reply by lukaszpolowczyk** [reply]
@eltigerchino I don't know. Should I know?

**Reply by teemingc** [reply]
Oh I'm so sorry. I tagged the wrong Lukas 😭 @Lms24 this is the correct one

**Reply by Lms24** [reply]
Hi, not that I'm aware of on first glance but I agree that this is far from ideal. I'll convert this thread to an issue for us to track. If anyone has ideas in the meantime, feel free to share them!

UPDATE: This is already tracked in https://github.com/getsentry/sentry-javascript/issues/16099. Let's continue the discussion over there!
